### PR TITLE
Error for using positional member that is hidden

### PR DESCRIPTION
**Pull request by jcouv**
Fixes https://github.com/dotnet/roslyn/issues/52630

Relates to test plan: https://github.com/dotnet/roslyn/issues/51199
Spec change: https://github.com/dotnet/csharplang/pull/4673

### CODE REVIEW

File: src/Compilers/CSharp/Portable/CSharpResources.resx [hunk 1]
```diff
@@ -6606,4 +6606,7 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
   <data name="ERR_InheritingFromRecordWithSealedToString" xml:space="preserve">
     <value>Inheriting from a record with a sealed 'Object.ToString' is not supported in C# {0}. Please use language version '{1}' or greater.</value>
   </data>
+  <data name="ERR_HiddenPositionalMember" xml:space="preserve">
+    <value>The positional member `{0}` found corresponding to this parameter is hidden.</value>
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/CSharpResources.resx:6610]: for consistency 

```suggestion
    <value>The positional member '{0}' found corresponding to this parameter is hidden.</value>
```

Comment (jcouv) [reply src/Compilers/CSharp/Portable/CSharpResources.resx]: Thanks. Fixed

File: docs/compilers/CSharp/Compiler Breaking Changes - post DotNet 5.md [hunk 2]
```diff
@@ -34,3 +34,17 @@
...
     }
     ```
+
+3. https://github.com/dotnet/roslyn/issues/52630 In C# 9 (.NET 5, Visual Studio 16.9), it is possible that a record uses a hidden member from a base type as a positional member. In Visual Studio 16.10, this is now an error:
+```csharp
+record Base
+{
+    public int I { get; init; }
+}
+record Derived(int I) // error: The positional member `Base.I` found corresponding to parameter `I` is hidden.
```
Comment (Youssef1313) [inline docs/compilers/CSharp/Compiler Breaking Changes - post DotNet 5.md:44]:
```suggestion
record Derived(int I) // error: The positional member 'Base.I' found corresponding to this parameter is hidden.
```

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordDeconstruct.cs [hunk 3]
```diff
@@ -59,6 +61,27 @@ protected override (TypeWithAnnotations ReturnType, ImmutableArray<ParameterSymb
...
+        protected override void MethodChecks(BindingDiagnosticBag diagnostics)
+        {
+            base.MethodChecks(diagnostics);
+
+            if (ParameterCount == _properties.Length)
+            {
+                for (int i = 0; i < _properties.Length; i++)
+                {
+                    var property = _properties[i];
+                    NamedTypeSymbol containingType = this.ContainingType;
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordDeconstruct.cs:73]: Could be moved outside the loop?

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordDeconstruct.cs [hunk 4]
```diff
@@ -59,6 +61,27 @@ protected override (TypeWithAnnotations ReturnType, ImmutableArray<ParameterSymb
 
         protected override int GetParameterCountFromSyntax() => _ctor.ParameterCount;
 
+        protected override void MethodChecks(BindingDiagnosticBag diagnostics)
+        {
+            base.MethodChecks(diagnostics);
+
+            if (ParameterCount == _properties.Length)
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordDeconstruct.cs:68]: - When can the this condition be false?


- What if the Deconstruct method is declared explicitly:

    ```csharp
    public record Base
    {
        public int I { get; set; } = 42;
        public Base(int ignored) { }
    
        public void Deconstruct(out string s, out string s2) { s = s2 = "";  }
    }
    public record C(int I) : Base(I)
    {
        public void I() { }
    
        public void Deconstruct(out string s, out string s2) { s = s2 = "";  }
    }
    ```

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordDeconstruct.cs:68]: - the condition is false when `addProperties` in `SourceMemberContainerSymbol` doesn't find a property for each parameter (we produce `ERR_BadRecordMemberForPositionalParameter` in that case).
- thanks. I need to think more about scenarios where `Deconstruct` is declared in source. We should produce a diagnostic too, but currently don't.

---
In reply to: [614614357](https://github.com/dotnet/roslyn/pull/52660#discussion_r614614357) [](ancestors = 614614357)

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordDeconstruct.cs [hunk 5]
```diff
@@ -59,6 +61,27 @@ protected override (TypeWithAnnotations ReturnType, ImmutableArray<ParameterSymb
...
+                for (int i = 0; i < _properties.Length; i++)
+                {
+                    var property = _properties[i];
+                    NamedTypeSymbol containingType = this.ContainingType;
+
+                    if (property.ContainingType != (object)containingType &&
+                        !containingType.GetMembers(property.Name).IsEmpty)
+                    {
+                        // The positional member was inherited but is hidden by a member of the current record type
+                        diagnostics.Add(ErrorCode.ERR_HiddenPositionalMember, _parameters.Parameters[i].Identifier.GetLocation(), property);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordDeconstruct.cs:79]: I think this isn't the right place to detect and report the error. I think this should be done when we are looking for positional members and regardless whether the "Deconstruct" is explicitly provided.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 6]
```diff
@@ -3745,23 +3745,23 @@ ImmutableArray<PropertySymbol> addProperties(ImmutableArray<ParameterSymbol> rec
...
-                        members.Add(property.BackingField);
-
-                        builder.AddInstanceInitializerForPositionalMembers(new FieldOrPropertyInitializer(property.BackingField, paramList.Parameters[param.Ordinal]));
-                        addedCount++;
-                    }
                 }
 
                 return existingOrAddedMembers.ToImmutableAndFree();
+
+                void addProperty(SynthesizedRecordPropertySymbol property, ParameterSymbol param)
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3752]: It looks like a refactoring is bundled with feature work. Consider reverting.

File: src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordDeconstruct.cs [hunk 7]
```diff
@@ -17,18 +15,22 @@ internal sealed class SynthesizedRecordDeconstruct : SynthesizedRecordOrdinaryMe
     {
         private readonly SynthesizedRecordConstructor _ctor;
         private readonly ImmutableArray<PropertySymbol> _properties;
+        private readonly ParameterListSyntax _parameters;
 
         public SynthesizedRecordDeconstruct(
             SourceMemberContainerTypeSymbol containingType,
             SynthesizedRecordConstructor ctor,
             ImmutableArray<PropertySymbol> properties,
+            ParameterListSyntax parameters,
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Synthesized/Records/SynthesizedRecordDeconstruct.cs:24]: Long lived symbols are not allowed to hold on to the syntax like that.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 8]
```diff
@@ -3410,8 +3410,14 @@ private void AddSynthesizedRecordMembersIfNecessary(MembersAndInitializersBuilde
             var memberSignatures = s_duplicateRecordMemberSignatureDictionary.Allocate();
             var membersSoFar = builder.GetNonTypeMembers(declaredMembersAndInitializers);
             var members = ArrayBuilder<Symbol>.GetInstance(membersSoFar.Count + 1);
+            var memberNames = PooledHashSet<string>.GetInstance();
             foreach (var member in membersSoFar)
             {
+                if (!memberNames.Contains(member.Name))
+                {
+                    memberNames.Add(member.Name);
+                }
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3419]: Doesn't `Add` method internally check whether the element is there or not?

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: It does. Thanks

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 9]
```diff
@@ -3735,6 +3739,11 @@ ImmutableArray<PropertySymbol> addProperties(ImmutableArray<ParameterSymbol> rec
                         {
                             // Deconstruct() is specified to simply assign from this property to the corresponding out parameter.
                             existingOrAddedMembers.Add(prop);
+
+                            if (isInherited)
+                            {
+                                validateNotHidden(prop, param);
```
Comment (Youssef1313) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3745]: If `prop` is abstract, it won't be validated, is that intended? i.e:

```csharp
abstract record Base
{
    public abstract int I { get; init; }
}

record Derived(int I) : Base
{
    public int I() { return 0; }
}
```

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3745]: In that scenario `Derived` defined a concrete property `I` and a method `I()` (that's an error). The positional member is `Derived.I` so we didn't pick a hidden member, so the new error doesn't come into play.
Added tests to illustrate. Thanks for the scenario.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 10]
```diff
@@ -3762,6 +3771,14 @@ void addProperty(SynthesizedRecordPropertySymbol property)
                 }
 
                 return existingOrAddedMembers.ToImmutableAndFree();
+
+                void validateNotHidden(Symbol symbol, ParameterSymbol param)
+                {
+                    if (memberNames.Contains(symbol.Name) || !this.GetTypeMembers(symbol.Name).IsEmpty)
```
Comment (jcouv) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3777]: :memo: I have to say I'm a bit worried about this. The next feature we add where it turns out that computing type-members depends on computing non-type-members will cause a cycle.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 8 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: The behavior of Add is to throw when the key is already present. Is that what we want? Consider testing a scenario where two fields with the same name are declared in the type.

Comment (Youssef1313) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: @RikkiGibson Isn't it supposed to return `false` when the item already exists?

Comment (RikkiGibson) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: Oops, sounds like I mixed up the behavior for Dictionary vs for Set. 😅

Comment (jcouv) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs]: I did the same :-P Had to check the docs ;-)

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs (hunk 10 above)
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3777]: >GetTypeMembers [](http://example.com/codeflow?start=67&length=14)

```GetTypeMembersDictionary().Contains```?

Comment (AlekseyTs) [reply src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3777]: >  I have to say I'm a bit worried about this. The next feature we add where it turns out that computing type-members depends on computing non-type-members will cause a cycle.

I don't think this adds any new possible exposure. We are binding signatures of members in ```AddSynthesizedRecordMembersIfNecessary``` and those can refere to types declared inside the record.

File: src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs [hunk 11]
```diff
@@ -3735,6 +3739,11 @@ ImmutableArray<PropertySymbol> addProperties(ImmutableArray<ParameterSymbol> rec
                         {
                             // Deconstruct() is specified to simply assign from this property to the corresponding out parameter.
                             existingOrAddedMembers.Add(prop);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:3741]: >existingOrAddedMembers [](http://example.com/codeflow?start=28&length=22)

 Consider not adding hidden members here

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 12]
```diff
@@ -28680,5 +28680,353 @@ public void SealedIncomplete()
...
+                // (7,21): error CS8913: The positional member 'Base.I' found corresponding to this parameter is hidden.
+                // public record C(int I) : Base(I)
+                Diagnostic(ErrorCode.ERR_HiddenPositionalMember, "I").WithArguments("Base.I").WithLocation(7, 21),
+                // (9,17): warning CS0108: 'C.I()' hides inherited member 'Base.I'. Use the new keyword if hiding was intended.
+                //     public void I() { }
+                Diagnostic(ErrorCode.WRN_NewRequired, "I").WithArguments("C.I()", "Base.I").WithLocation(9, 17)
+            };
+
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular9);
+            comp.VerifyDiagnostics(expected);
```
Comment (AlekseyTs) [inline src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:28709]: >VerifyDiagnostics [](http://example.com/codeflow?start=17&length=17)

```VerifyEmitDiagnostics```? For other new error scenarios too.

**Comment by AlekseyTs** [comment]
Done with review pass (commit 9)

File: src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs [hunk 13]
```diff
@@ -28680,5 +28680,353 @@ public void SealedIncomplete()
...
+abstract record Base
+{
+    public abstract int I { get; init; }
+}
+abstract record Derived(int I) : Base
+{
+    public int I() { return 0; }
+}
+";
+            var comp = CreateCompilation(new[] { source, IsExternalInitTypeDefinition }, parseOptions: TestOptions.Regular9);
```
[the remark this answers is not in the document]
Comment (jcouv) [reply src/Compilers/CSharp/Test/Semantic/Semantics/RecordTests.cs:29021]: :memo: I'm fine either way. I also changed the default for the entire file then.

---
In reply to: [616038511](https://github.com/dotnet/roslyn/pull/52660#discussion_r616038511) [](ancestors = 616038511)

**Review by AlekseyTs — APPROVED** [review]
LGTM (commit 11)

**Review by RikkiGibson — APPROVED** [review]

**Pull request merged**